### PR TITLE
Lab1

### LINKED ISSUE #162: Lab 1 - fibonacci series
- [ ] Create an executable go program in directory 01_fib/CameronSchafer
- [ ] Write a function that prints the first n Fibonacci numbers
```
func fib(n int)
```
- [ ] Call fib(7) in main to print
```
1
1
2
3
5
8
13
```
- [ ] Bonus points: For negative n print Negafibonacci numbers.

### PR DESCRIPTION
**Pull request by CameronSchafer**
Fixes #162

Review of colleague's PR #133

#### Changes proposed in this PR:
- Implement Lab1 Fibonacci numbers
- Implement Lab1 NegaFibonacci numbers
- Test Lab1

### CODE REVIEW

File: 01_fib/CameronSchafer/fib.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+//base fibonacci function
+func fib(n int) {
+	sign, n1, n2 := 1, 0, 1
+	//check if use nega or normal fib sequence
+	if n < 0 {
+		sign = -1
+	}
+	//loop until all numbers have been printed
+	for i := 0; i < n*sign; i++ {
```
Comment (nicholascross) [inline 01_fib/CameronSchafer/fib.go]: We could potentially optimise this by removing the multiplication everytime the the condition is evaluated.  Calculating the sequence length once in advance given the value `n*sign` does not change during iteration.

File: 01_fib/CameronSchafer/fib.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+		if i%2 == 0 || sign > 0 {
+			//will get the negafib number if the sign is -1
+			n1, n2 = n2, n1+n2*sign
```
Comment (nicholascross) [inline 01_fib/CameronSchafer/fib.go:24]: For my own benefit, can you tell me does this evaluate everything on the right of `=` prior to assignment?  I guess it must.

File: 01_fib/CameronSchafer/fib_test.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+-3
+5
+-8
+13
+`),
+		},
+	}
+
+	for _, test := range testCases {
+		test := test
```
Comment (nicholascross) [inline 01_fib/CameronSchafer/fib_test.go:68]: Why do we need to do this?

File: 01_fib/CameronSchafer/fib.go (hunk 2 above)
Comment (nicholascross) [reply 01_fib/CameronSchafer/fib.go:24]: Thanks I found it does do that, this has helped me simplify my second lab 👍

**Comment by nicholascross** [comment]
This looks good to me the tests are passing and seem to cover most interesting edge cases except maybe `0`

I had a suggestion and question for the author but no required changes.

File: 01_fib/CameronSchafer/fib.go (hunk 2 above)
Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib.go:24]: Yes it does. This assignment is shorthand for
```
n1 = n2
n2 = n1 + (n2 * sign)
```

File: 01_fib/CameronSchafer/fib_test.go (hunk 3 above)
Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib_test.go:68]: Without this assignment, test was not accessible within the for loop.

File: 01_fib/CameronSchafer/fib.go (hunk 1 above)
Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib.go]: Thank you, I will try change this up now.

Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib.go]: @nicholascross I updated the code as per your recommendation.

File: 01_fib/CameronSchafer/fib.go (hunk 2 above)
Comment (nicholascross) [reply 01_fib/CameronSchafer/fib.go:24]: From my experiments it is more closely equivalent to

```
x1 := n1
x2 := n2

n1 = x2 
n2 = x1 + (x2 * sign)
```
So the values on the right hand side are always the original values before any assignment occurs.

So you can swap variable values like this

```
n1, n2 = n2, n1
```

You can see if it translated as seperate statements you would end up with the same value in `n1` and `n2`

File: 01_fib/CameronSchafer/fib_test.go (hunk 3 above)
Comment (nicholascross) [reply 01_fib/CameronSchafer/fib_test.go:68]: Is it something to do with the anonymous function/closure?

Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib_test.go:68]: To be honest, I'm not exactly sure. I think it has something to do with the way I was trying to access the struct array. I will look into it some more and get back to you.

Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib_test.go:68]: It is a linting error that pops up if I don't do that assignment:
`Using the variable on range scope 'test' in function literal (scopelint)`

So I feel it is safe to just do the assignment to remove the error.

File: 01_fib/CameronSchafer/fib.go (hunk 2 above)
Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib.go:24]: Yeah that makes sense, it is a really useful feature of Go.

File: 01_fib/CameronSchafer/fib_test.go (hunk 3 above)
Comment (nicholascross) [reply 01_fib/CameronSchafer/fib_test.go:68]: @CameronSchafer I think I found the [answer](https://stackoverflow.com/questions/26692844/captured-closure-for-loop-variable-in-go)

When your function literal captures the `test` variable it is capturing a `reference to the value` not the `value itself`.  So if the functions are executed outside of the for loop then they will have the last value of the range not the value at the time of iteration.

So you have done the right thing here 👍

Comment (CameronSchafer) [reply 01_fib/CameronSchafer/fib_test.go:68]: @nicholascross Oh Ok. That makes sense now. Thank you

File: 01_fib/CameronSchafer/fib.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+//base fibonacci function
+func fib(n int) {
+	sign, n1, n2 := 1, 0, 1
+	//check if use nega or normal fib sequence
+	if n < 0 {
+		sign = -1
+	}
+
+	n *= sign
```
Comment (camscale) [inline 01_fib/CameronSchafer/fib.go:19]: minor suggestion: you could move this inside the `n < 0` block above since it only needs to be done for negative `n`.

File: 01_fib/CameronSchafer/fib.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	//loop until all numbers have been printed
+	for i := 0; i < n; i++ {
+		if i%2 == 0 || sign > 0 {
+			//will get the negafib number if the sign is -1
+			n1, n2 = n2, n1+n2*sign
+		} else {
+			//will account for the negative number during the negafib sequence
+			n1, n2 = n2, n1-n2
```
Comment (camscale) [inline 01_fib/CameronSchafer/fib.go:27]: Since this branch of the `if` statement is only taken when `sign == -1`, and `n1+n2*-1` is the same as `n1-n2`, the two branches of the `if` statement are functionally identical. You could remove the `if` statement entirely and just leave the first branch.

File: 01_fib/CameronSchafer/fib.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+func fib(n int) {
+	sign, n1, n2 := 1, 0, 1
+	//check if use nega or normal fib sequence
+	if n < 0 {
+		sign = -1
+	}
+
+	n *= sign
+	//loop until all numbers have been printed
+	for i := 0; i < n; i++ {
```
Comment (camscale) [inline 01_fib/CameronSchafer/fib.go:21]: minor suggestion: count from 1 to n, instead of 0 to n-1 (i.e. `for i := 1; i <= n; i++`) because we are computing the nth fibonacci number, not the nth-1 fibonacci number.

File: 01_fib/CameronSchafer/fib_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+			expected: strconv.Quote(`1
+-1
+2
+-3
+5
+-8
+13
+`),
+		},
+	}
```
Comment (camscale) [inline 01_fib/CameronSchafer/fib_test.go:65]: I would add a `fib(0)` test case too as it is on the boundary.

**Review by camscale — APPROVED** [review]
I'm approving this with suggestions as the suggestions all relate to software engineering rather than Go specifically. From the perspective of Go, this code is great and shows some good idiomatic use of Go.

A minor nitpick, but I'll also note that your comment style is at odds with most other Go code in that comments start with slash-slash-space (`// `), rather than just slash-slash (`//`).
